drivers/eeprom/eeprom_st25dv.c

```diff
@@ -0,0 +1,970 @@
+	 * completed.
+	 */
+	timeout = k_uptime_get() + EEPROM_ST25DV_WRITE_TIMEOUT;
+	while (1) {
+		int64_t now = k_uptime_get();
+
+		ret = i2c_write(config->bus.bus, buffer, count + 2, addr);
+		if (!ret || now > timeout) {
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	}
+	return ret;
+}
+
+static int _st25dv_write_user(const struct eeprom_st25dv_config *config, off_t offset,
+			      const uint8_t *pdata, uint32_t count)
+{
+	return _st25dv_write(config, config->addr, offset, pdata, count);
+}
+
+static int _st25dv_read_user(const struct eeprom_st25dv_config *config, off_t offset,
+			     uint8_t *pdata, uint32_t count)
+{
+	return _st25dv_read(config, config->addr, offset, pdata, count);
+}
+
+static int _st25dv_read_conf(const struct eeprom_st25dv_config *config, off_t offset,
+			     uint8_t *pdata, uint32_t count)
+{
+	return _st25dv_read(config, config->addr | 4, offset, pdata, count);
+}
+
+static int _st25dv_write_conf(const struct eeprom_st25dv_config *config, off_t offset,
+			      uint8_t *pdata, uint32_t count)
+{
+	return _st25dv_write(config, config->addr | 4, offset, pdata, count);
+}
+
+static int eeprom_st25dv_read(const struct device *dev, off_t offset, void *buf, size_t len)
+{
+	const struct eeprom_st25dv_config *config = dev->config;
+	struct eeprom_st25dv_data *data = dev->data;
+	int ret;
+
+	if (!len) {
+		return 0;
+	}
+
+	if ((offset + len) > config->size) {
+		LOG_WRN("attempt to read past device boundary");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	ret = _st25dv_read_user(config, offset, buf, len);
+	k_mutex_unlock(&data->lock);
+
+	return ret;
+}
+
+static int eeprom_st25dv_write(const struct device *dev, off_t offset, const void *buf, size_t len)
+{
+	const struct eeprom_st25dv_config *config = dev->config;
+	struct eeprom_st25dv_data *data = dev->data;
+	int ret;
+
+	if (!len) {
+		return 0;
+	}
+
+	if ((offset + len) > config->size) {
+		LOG_WRN("attempt to write past device boundary");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+	ret = _st25dv_write_user(config, offset, buf, len);
+	k_mutex_unlock(&data->lock);
+
+	return ret;
+}
+
+static size_t eeprom_st25dv_size(const struct device *dev)
+{
+	const struct eeprom_st25dv_config *config = dev->config;
+
+	return config->size;
+}
+
+static int eeprom_st25dv_init(const struct device *dev)
+{
+	const struct eeprom_st25dv_config *config = dev->config;
+	struct eeprom_st25dv_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("parent bus device not ready");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+/**
+ * @brief  Read ENDA1 register
+ * @param[in] config ST25DV config structure
+ * @param[out] value data pointer to store register content
+ * @return 0 in case of success, an error code otherwise
+ */
+static int _get_enda1(const struct eeprom_st25dv_config *config, uint8_t *const value)
+{
+	if (value == NULL) {
+		return -EINVAL;
+	}
```

Comment:
Asserts would probably be fine here.

Reply:
Updated to use assert in every NULL check for parameters of function. Is assert() from assert.h the right way to do it ?